// myorm_pg/lib/implementations/PGSetHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class PGSetHelper {
    static InjectORMData(set, key, value) {
        set[key] = value;
        return set;
    }
    static ExtractORMData(set, key) {
        var _a;
        return (_a = set[key]) !== null && _a !== void 0 ? _a : "";
    }
    static InjectSQL(set, sql) {
        return PGSetHelper.InjectORMData(set, PGSetHelper._sqlKey, sql);
    }
    static InjectWhere(set, where) {
        return PGSetHelper.InjectORMData(set, PGSetHelper._whereKey, where);
    }
    static InjectJoin(set, join) {
        return PGSetHelper.InjectORMData(set, PGSetHelper._joinKey, join);
    }
    static InjectOn(set, on) {
        return PGSetHelper.InjectORMData(set, PGSetHelper._onKey, on);
    }
    static ExtractJoinData(set) {
        return PGSetHelper.ExtractORMData(set, PGSetHelper._joinKey);
    }
    static ExtractOnData(set) {
        return PGSetHelper.ExtractORMData(set, PGSetHelper._onKey);
    }
    static ExtractWhereData(set) {
        return PGSetHelper.ExtractORMData(set, PGSetHelper._whereKey);
    }
    static ExtractSQLData(set) {
        return PGSetHelper.ExtractORMData(set, PGSetHelper._sqlKey);
    }
    static CleanORMData(set) {
        PGSetHelper.InjectJoin(set, "");
        PGSetHelper.InjectSQL(set, "");
        PGSetHelper.InjectOn(set, "");
        PGSetHelper.InjectSQL(set, "");
    }
}
PGSetHelper._sqlKey = "__orm__sql__";
PGSetHelper._joinKey = "__orm__join__";
PGSetHelper._onKey = "__orm__on__";
PGSetHelper._whereKey = "__orm__where__";
exports.default = PGSetHelper;
